myhpom/models/document.py

```diff
@@ -152,6 +211,224 @@ def authorized_client_ip(self, ip_address):
+            "units": [unit_post_data],
+        }
+        return data
+
+    def save_response_data(self, response_content):
+        """update the response_data field and the other fields that are extracted from it.
+        """
+        # we want to save the response_content to self.response_data no matter what.
+        self.response_content = response_content
+        self.save()
+
+        # now we try to unpack the response_content on the assumption that it is json.
+        try:
+            data = json.loads(response_content)  # throws an error if not json
+            if 'id' in data:
+                self.run_id = data['id']
+            if 'status' in data:
+                self.status = data['status']
+            if 'created_at' in data:
+                self.created_at = data['created_at']
+            if 'processed_at' in data:
+                self.processed_at = data['processed_at']
+        except ValueError:
+            self.status = self.STATUS_ERROR
+            self.save()
+            raise
+
+        self.save()
+
+    def output(self):
+        """
+        Returns dictionary of 'output' from run if is processed.
+
+        Otherwise if the run failed to process returns None or return non-JSON
+        return None.
+        """
+        if self.status != CloudFactoryDocumentRun.STATUS_PROCESSED:
+            return None
+
+        try:
+            data = json.loads(self.response_content)
+            if 'units' in data:
+                units = data['units']
+                if len(units) > 0 and 'output' in units[0]:
+                    return units[0]['output']
+        except ValueError:
+            # fall thru to failure
+            pass
+
+        return None
+
+    def passed(self):
+        """
+        Returns True when this run processed successfully, and all the
+        outputs are either true or na.
+        """
+        if self.status != CloudFactoryDocumentRun.STATUS_PROCESSED:
+            return False
+
+        try:
+            data = json.loads(self.response_content)
+            if 'units' not in data.keys():
+                return False
+            units = data['units']
+            if len(units) == 0 or 'output' not in units[0].keys():
+                return False
+
+            # If all the values are either true or not applicable then this
+            # would be considered a successful run:
+            #
+            # Note that we consider only the first unit of work - which at the
+            # moment is all we create.
+            #
+            # The output must also have at least the required keys mentioned in
+            # our integration doc.
+            output = units[0]['output']
+            all_true_or_na = set(output.values()) <= self.YES_OR_NA
+            has_required_keys = self.REQUIRED_OUTPUT_KEYS <= set(output.keys())
+            return all_true_or_na and has_required_keys
+        except ValueError:
+            return False
```

Comment:
add a KeyError block here as well? 

With both of these, we have a situation where CloudFactory has returned a response without output. This seems to be an error condition that should raise our attention, set the run to STATUS_ERROR, and show the user the "In verification" message.

This suggests that we should audit the callback a little more so that we never get to the point of having a STATUS_PROCESSED that lacks an ['output'] key.